### PR TITLE
Detect missing block for bitcoin

### PR DESCRIPTION
**Pull request by bonomat**
This is the first step towards lite clients.
I kept the change as small as possible, so this PR introduces: 

* `Bitcoin` and `Ethereum`-chain in https://github.com/comit-network/comit-rs/blob/a32b6cbf904e68c6bec8dad856fc6a24c98ff763/btsieve/src/lib.rs#L58-L60
* `BlockchainDAG` data-structure to model the blockchain as a _Directed Acylic Graph_ https://github.com/comit-network/comit-rs/blob/a32b6cbf904e68c6bec8dad856fc6a24c98ff763/btsieve/src/lib.rs#L44
* and an implementation to store incoming blocks and check for missed blocks in: https://github.com/comit-network/comit-rs/blob/a32b6cbf904e68c6bec8dad856fc6a24c98ff763/btsieve/src/bitcoin/mod.rs#L12-L43

--> The idea is to add the same abstraction for Ethereum next (TODO create new ticket) and then update the way of how blocks are being processed.

### CODE REVIEW

File: btsieve/src/bitcoin/mod.rs [hunk 1]
```diff
@@ -6,3 +6,161 @@ pub use self::{
...
+use bitcoin_support::{MinedBlock, Sha256dHash};
+
+impl AddBlock<MinedBlock> for Bitcoin {
+    fn add_block(&mut self, block: MinedBlock) {
+        if self.0.nodes.contains(&block) {
+            return log::warn!("Block already known {:?} ", block);
+        }
+        match self.find_predecessor(&block) {
+            Some(prev) => {
+                self.0.vertices.push((
```
Comment (D4nte) [inline btsieve/src/bitcoin/mod.rs]: You went further than needed, the storage of links between blocks is out of scope :)

File: btsieve/src/lib.rs [hunk 2]
```diff
@@ -38,3 +40,27 @@ impl PartialOrd for QueryMatch {
         self.0.partial_cmp(&other.0)
     }
 }
+
+struct BlockchainDAG<T, V> {
```
Comment (D4nte) [inline btsieve/src/lib.rs]: The Blockchain DAG is not needed for this PR :)

Comment (bonomat) [reply btsieve/src/lib.rs]: I know, but otherwise this would have been too easy :D

File: btsieve/src/bitcoin/mod.rs [hunk 3]
```diff
@@ -6,3 +6,161 @@ pub use self::{
...
+                .merkle_root
+                .eq(&block.block.header.prev_blockhash)
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::BlockchainDAG;
```
Comment (tcharding) [inline btsieve/src/bitcoin/mod.rs]: In line with Rust convention can we name this `BlockchainDag` please.  (Acronyms not capitalized, source: @luckysori)

File: btsieve/src/bitcoin/mod.rs [hunk 4]
```diff
@@ -6,3 +6,161 @@ pub use self::{
...
+            .unwrap(),
+            Sha256dHash::from_hex(
+                "0000000000000000000000000000000000000000000000000000000000000002",
+            )
+            .unwrap(),
+            1,
+        );
+
+        assert_that(&bitcoin_chain.size()).is_equal_to(&0);
+        bitcoin_chain.add_block(block);
```
Comment (tcharding) [inline btsieve/src/bitcoin/mod.rs]: Perhaps another assert statement here `assert_that(&bitcoin_chain.size()).is_equal_to(&1);`

File: btsieve/src/lib.rs [hunk 5]
```diff
@@ -38,3 +40,27 @@ impl PartialOrd for QueryMatch {
...
+            vertices: Vec::new(),
+        })
+    }
+}
+
+struct Bitcoin(BlockchainDAG<MinedBlock, (Sha256dHash, Sha256dHash)>);
+
+struct Ethereum;
+
+trait AddBlock<T> {
```
Comment (tcharding) [inline btsieve/src/lib.rs]: I do not know typical Rust naming for traits but this name does not seem to match the trait functions. Suggest `Blockchain`.

Comment (tcharding) [reply btsieve/src/lib.rs]: Also perhaps `default` should be in the trait (if that is how Rust likes it)?

File: btsieve/src/bitcoin/mod.rs (hunk 4 above)
Comment (bonomat) [reply btsieve/src/bitcoin/mod.rs]: Mhmm... must have deleted that one :D

File: btsieve/src/lib.rs (hunk 5 above)
Comment (bonomat) [reply btsieve/src/lib.rs]: True, at the beginning there was only one function --> I'll update it.

**Pull request merged**